src/main/java/org/breedinginsight/brapi/v2/BrAPITrialsController.java

```diff
@@ -57,9 +95,22 @@ public HttpResponse<Response<DataResponse<List<BrAPITrial>>>> getExperiments(
             @PathVariable("programId") UUID programId,
             @QueryValue @QueryValid(using = ExperimentQueryMapper.class) @Valid ExperimentQuery queryParams) {
         try {
+            List<BrAPITrial> experiments = new ArrayList<>();
             log.debug("fetching trials for program: " + programId);
 
-            List<BrAPITrial> experiments = experimentService.getExperiments(programId).stream().peek(this::setDbIds).collect(Collectors.toList());
+            Optional<ProgramUser> experimentalCollaborator = programUserService.getIfExperimentalCollaborator(programId, securityService.getUser().getId());
+            // If the program user is an experimental collaborator, filter results.
+            if (experimentalCollaborator.isPresent()) {
+                Optional<Program> program = programService.getById(programId);
+                if (program.isEmpty()) {
+                    return HttpResponse.notFound();
+                }
+                List<UUID> experimentIds = experimentalCollaboratorService.getAuthorizedExperimentIds(experimentalCollaborator.get().getId());
+                experiments = experimentService.getTrialsByExperimentIds(program.get(), experimentIds).stream().peek(this::setDbIds).collect(Collectors.toList());
+            } else {
+                experiments = experimentService.getExperiments(programId).stream().peek(this::setDbIds).collect(Collectors.toList());
+            }
+
```

Comment:
```suggestion
          Optional<HttpResponse> httpResponse = programUserService.getIfExperimentalCollaborator(programId, securityService.getUser().getId())
        .flatMap(experimentalCollaborator ->
                programService.getById(programId)
                        .map(program ->
                                experimentalCollaboratorService.getAuthorizedExperimentIds(experimentalCollaborator.getId())
                                        .stream()
                                        .flatMap(experimentId -> experimentService.getTrialsByExperimentIds(program, List.of(experimentId)).stream())
                                        .peek(this::setDbIds)
                                        .collect(Collectors.toList())))
        .or(() ->
                Optional.of(experimentService.getExperiments(programId)
                        .stream()
                        .peek(this::setDbIds)
                        .collect(Collectors.toList()))
                .filter(experiments -> !experiments.isEmpty())
                .map(HttpResponse::ok)
                .orElse(HttpResponse.notFound());

return httpResponse.orElse(HttpResponse.notFound());

```

Reply:
See my comment on the BrAPIStudiesController.java suggestion.